chain-impl-mockchain/doc/format.md

```diff
@@ -2,304 +2,330 @@
+Inputs/Outputs is in the following format:
 
-    TokenTransfer<PAYLOAD = ()> | Witnesses
+```
+                                        <--- N-INPS times --->
+IOS = N-INPS (1 byte) | N-OUTS (1 byte) | INPUT | .. | INPUT | OUTPUT | .. | OUTPUT 
+                                                             <--- N-OUTS times --->
 
-TODO:
+INPUT = INPUT-INDEX (1 byte) | INPUT-VALUE (8 bytes) | INPUT-REFERENCE 
 
-* Multisig
-* Fees
+INPUT-REFERENCE = ACCOUNT-PUBKEY (32 bytes) || FRAGMENT-ID (32 bytes)
 
-## Type 2: OwnerStakeDelegation
+OUTPUT = OUTPUT-ADDRESS (33 or 65 bytes) | OUTPUT-VALUE (8 bytes)
+```
 
-    TokenTransfer<PAYLOAD = OwnerStakeDelegation> | Witnesses
+##### Remarks
 
-    OwnerStakeDelegation = StakePoolId
+1. 256 inputs maximum.
 
-## Type 3: Certificate
+2. 255 outputs maximum, `0xff` is reserved. 
 
-Certificate is the composition of the TokenTransfer structure where PAYLOAD is the certificate data, and then the witnesses. Effectively:
+3. All `INPUT` are therefore `41` bytes.
 
-    TokenTransfer<PAYLOAD = CERTIFICATE> | Witnesses
+4. A special `INPUT-INDEX` value of `0xff` specifies an account spending.
 
-Known Certificate types:
+#### Witnesses
 
-* Staking declaration: declare a staking key + account public information
-* Stake pool registration: declare the VRF/KES key for a node.
-* Delegation: contains a link from staking to stake pool.
+To authenticate the `PAYLOAD` and the IOs, we add witnesses with a 1-to-1 mapping
+with inputs. The serialized sequence of inputs, is directly linked with the
+serialized sequence of witnesses.
 
-Content:
+Fundamentally the witness is about signing a message and generating/revealing
+cryptographic material to approve unequivocally the content.
 
-* PublicKey
-* Signature of the witness with the private key associated to the revealed PublicKey
+There's currently 3 differents types of witness supported:
+
+1. Old utxo scheme: an extended public key, followed by a ED25519 signature
+1. Utxo scheme: a ED25519 signature
+1. Account scheme: a counter and an ED25519 signature
 
-## Type 4: Update Proposal
+With the following serialization:
 
-Update proposal messages propose new values for blockchain
-settings. These can subsequently be voted on. They have the following
-form:
+```
+            <----------------- N-INPS times ----------------->
+WITNESSES = WITNESS{WITNESS-TYPE} | .. | WITNESS{WITNESS-TYPE}
 
-    Proposal | ProposerId | Signature
+WITNESS{OLD-UTXO} = WITNESS-TYPE (1 byte) | EXTENDED-PUBKEY (64 bytes) | WITNESS-SIGNATURE (64 bytes)
 
-where `ProposerId` is a ed25519 extended public key, and `Signature`
-is a signature by the corresponding private key over the string
-`Proposal | ProposerId`.
+WITNESS{UTXO} = WITNESS-TYPE (1 byte) | WITNESS-SIGNATURE (64 bytes)
 
-`Proposal` has the following format:
+WITNESS{ACCOUNT} = WITNESS-ACCOUNT-COUNTER (4 bytes) | WITNESS-SIGNATURE (64 bytes)
```

Comment:
```suggestion
WITNESS{ACCOUNT} = WITNESS-SIGNATURE (64 bytes)
```
As per discussion on Slack, the account counter will actually not show up in the witness.

Reply:
Good catch. The counter seems however part of the `WITNESS-SIGNATURE-DATA` 

https://github.com/input-output-hk/chain-libs/blob/master/chain-impl-mockchain/src/transaction/witness.rs#L82-L85

I'll make the corresponding updates.

Reply:
Done: https://github.com/input-output-hk/chain-libs/compare/4c963df8a5be730aaca5bd0b3d5f9278ec988ee3..d0af1a03583980702bae16625c4920f8fbfbe377#diff-8f0a0fbb888b790f5dbc6717bcca5d8fR316